### PR TITLE
Extended Euclid to find gen class

### PR DESCRIPTION
**Pull request by inthar-raven**

### CODE REVIEW

**Comment by inthar-raven** [comment]
The algorithm is described in Definition 3.1 (page 6) in this paper: https://arxiv.org/abs/2012.15818

File: src/index.ts [hunk 1]
```diff
@@ -82,6 +82,7 @@ function bjorklund(subsequences: any[][]) {
  * @returns The array of evenly mixed booleans
  */
 export function euclid(numberOfTrue: number, numberOfFalse: number): boolean[] {
+  // return bresenham(numberOfTrue, numberOfFalse, true, false);
```
Comment (frostburn) [inline src/index.ts:85]: You forgot to actually use your new method.

File: src/index.ts [hunk 2]
```diff
@@ -151,15 +195,9 @@ function mosGeneratorMonzo(l: number, s: number): [number, number] {
...
-  for (let i = 1; i < t; ++i) {
-    if ((s * i) % t === 1) {
-      brightGeneratorSteps = i;
-      break;
-    }
-  }
+  const brightGeneratorSteps = modInv(s, t);
 
-  // Obtain some MOS pattern
+  // Obtain the brightest mode of the MOS pattern
```
Comment (frostburn) [inline src/index.ts:200]: If brightness is quaranteed, remove the dark fallback.

**Comment by frostburn** [comment]
This makes Björklund an orphan so it could be moved to `xen-dev-utils` so that it won't get lost. I can handle this piece of housekeeping after completing the review process.

**Comment by inthar-raven** [comment]
I want to first get the extended Euclidean algorithm part to work, and then we'll see if my Bresenham function or the existing Bjorklund function runs faster or if we can optimize the Bjorklund function.

**Comment by frostburn** [comment]
> I want to first get the extended Euclidean algorithm part to work, and then we'll see if my Bresenham function or the existing Bjorklund function runs faster or if we can optimize the Bjorklund function.

See how `xen-dev-utils` does benchmarking. We don't need to integrate this all the way to `scale-workshop` yet to have confidence in the speed.

File: src/index.ts [hunk 3]
```diff
@@ -111,6 +149,21 @@ const BRIGHT_GENERATORS: {[key: string]: [number, number]} = {
   '7,5': [3, 2],
 };
 
+/**
+ * Find the modular inverse of a mod b, provided gcd(a,b) == 1.
+ */
+function modInv(a: number, b: number): number {
+  const ee = extendedEuclid(a, b);
```
Comment (frostburn) [inline src/index.ts:156]: Just `const {gcd, coefA}  = extendedEuclid(a, b);`

File: src/index.ts (hunk 1 above)
Comment (frostburn) [reply src/index.ts:92]: Normally you shouldn't add commented pieces of code, but I guess this documents how to do the benchmarking so it's fine for now.

**Comment by frostburn** [comment]
`bresenham` is completely unreferenced now. Can you add tests to verify that it works (also documenting its use in the process)?

File: src/index.ts [hunk 4]
```diff
@@ -92,6 +100,36 @@ export function euclid(numberOfTrue: number, numberOfFalse: number): boolean[] {
...
+  const d = gcd(a, b);
+  if (d === 1) {
+    const result: boolean[] = [];
+    // `xHere` = current number of `first`, `yHere` = current number of `second`; start at (0, 0).
+    let [xHere, yHere] = [0, 0];
+    while (xHere < a || yHere < b) {
+      // If going north (taking a (0, 1) step) doesn't lead to going north of the line y = b/a*x,
+      if (a * (yHere + 1) <= b * xHere) {
+        // append `second` to `resultScale` and update the current location.
+        result.push(second);
```
Comment (frostburn) [inline src/index.ts:116]: Pushing `any` into an array of `boolean`s. Check the types.

File: src/index.ts [hunk 5]
```diff
@@ -92,6 +100,36 @@ export function euclid(numberOfTrue: number, numberOfFalse: number): boolean[] {
   return bjorklund(subsequences).reduce((a, b) => a.concat(b), []);
 }
 
+// This algorithm, a variant of the Bresenham line algorithm, returns the "brightest mode" of
+// the "scale" where `first` is treated as larger than `second`.
+// It's based on following the closest approximation of the line y = b/a*x that is strictly below the line.
+function bresenham(a: number, b: number, first: any, second: any): any[] {
```
Comment (frostburn) [inline src/index.ts:106]: I feel this is overly generic. Just push `true` in place of `first` and `false` in place of `second`. Remove these parameters.

File: src/index.ts [hunk 6]
```diff
@@ -92,6 +100,36 @@ export function euclid(numberOfTrue: number, numberOfFalse: number): boolean[] {
...
+  const d = gcd(a, b);
+  if (d === 1) {
+    const result: boolean[] = [];
+    // `xHere` = current number of `first`, `yHere` = current number of `second`; start at (0, 0).
+    let [xHere, yHere] = [0, 0];
+    while (xHere < a || yHere < b) {
```
Comment (frostburn) [inline src/index.ts:112]: `while` loops can be tricky. Remember to add tests for the cases where `a` or `b` is zero, negative or `NaN` in case of user mis-input. Don't want to hang the app.

Comment (frostburn) [reply src/index.ts:112]: I guess `a` and `b` should be positive non-zero integers based on the context. This whole package could use some input validation... That's on me.

**Comment by frostburn** [comment]
Issues minor. Will fix on `main`.
